File: Python_Version/FlameDetection.py
# import the necessary packages
import numpy as np
import cv2
import math

import sys
sys.path.append('./../CONSTANTS')
import CONSTANTS as CON

def Motion_Detaction(prev , next):
	
	flow = cv2.calcOpticalFlowFarneback(prev,next,None, 0.5,3,15,3,5,1.2,0)
	
	mag,ang = cv2.cartToPolar(flow[...,0],flow[...,1])

	hsv[...,0] = ang*180/np.pi/2
	hsv[...,2] = cv2.normalize(mag,None,0,255,cv2.NORM_MINMAX)
	
	bgr = cv2.cvtColor(hsv,cv2.COLOR_HSV2BGR)
	x = np.subtract(prev_bgr,bgr)
	r = x[:,:,0].flatten()
	g = x[:,:,1].flatten()
	b = x[:,:,2].flatten()
	
	# if 33 % of pixels any channel change their value by more than abs(10)
	thresh_hold = (x.shape[0] * x.shape[1])//3
	
	t1 = np.count_nonzero(r > 10)
	t2 = np.count_nonzero(g > 10)
	t3 = np.count_nonzero(b > 10)
	
	if(t1 >= thresh_hold or t2 >= thresh_hold or t3 >= thresh_hold):
		return True, bgr
	#print(t1,t2,t3)
	return False , bgr

	
def HSI_Color_Format(img):
	for i in range(img.shape[1]):
		for j in range(img.shape[0]):
			b , g, r = img[j,i]
			sum_bgr = int(int(b)+int(g)+int(r))
			I = sum_bgr/3
			S = 1 - 3*(min([r,g,b])/sum_bgr)
			if(S < 0.00001):
				S = 0
			elif(S > 0.99999):
				S = 1
			rb = int(int(r) - int(b))
			rg = int(int(r) - int(g))
			denom = math.sqrt(( rg * rg) + (rb*(int(g) - int(b))))
			if( S != 0 and denom!=0):
				H = 0.5 * (rg - rb) / math.sqrt(( rg * rg) + (rb*(int(g) - int(b))))
				H = math.acos(H)
				if( b <= g):
					H =H
				else:
					H = 360 - H
			else:
				H = 180
			S = S * 100
			img[j,i] = [I,S,H]
	count = 0
	for i in range(img.shape[1]):
		for j in range(img.shape[0]):
			I ,S , H = img[j,i]
			#Tune values of H,S,I as per demand 
			if (~((H<=120) and (S<=50) and (I>=180))):
				img[j,i] = [0,0,0]
				count+=1
				
	size = (img.shape[1] * img.shape[0])
	flame_per = 100 - (count*100)/size
	
	if flame_per < CON.CONST_FLAME_THRESHOLD:
		#print(flame_per)
		return False
	return True
	
def Flame_Area_Detection(gray):
	thresh = 127
	im_bw = cv2.threshold(gray, thresh, 255, cv2.THRESH_BINARY)[1]
	
	img_linear = im_bw.flatten()
	black_pixel_count = np.count_nonzero(img_linear == 0)
	white_pixel_count = np.count_nonzero(img_linear == 255)
	per_b_pixel = (white_pixel_count/len(img_linear)) * 100
	
	if(per_b_pixel < CON.CONST_AREA_THRESHOLD):
		#print(per_b_pixel)
		return False
	return True

	


cap = cv2.VideoCapture('./../Data/flame.mkv')
ret ,img =  cap.read()

clone = img.copy()
flame_img = clone[CON.CONST_UPPER_LEFT_Y:CON.CONST_BOTTOM_RIGHT_Y , CON.CONST_UPPER_LEFT_X:CON.CONST_BOTTOM_RIGHT_X]

prev = cv2.cvtColor(flame_img , cv2.COLOR_BGR2GRAY)


hsv = np.zeros_like(flame_img)
hsv[...,1] = 255
prev_bgr =hsv


while(True):
	ret, frame = cap.read()
	
	clone = frame.copy()
	next_frame = clone[CON.CONST_UPPER_LEFT_Y:CON.CONST_BOTTOM_RIGHT_Y , CON.CONST_UPPER_LEFT_X:CON.CONST_BOTTOM_RIGHT_X]
	next = cv2.cvtColor(next_frame, cv2.COLOR_BGR2GRAY)
	
	r1 , prev_bgr = Motion_Detaction(prev , next)
	r2 = HSI_Color_Format(next_frame)
	r3 = Flame_Area_Detection(next)

	if(r1 and r2 and r3):
		cv2.rectangle(frame, (CON.CONST_UPPER_LEFT_X, CON.CONST_UPPER_LEFT_Y), (CON.CONST_BOTTOM_RIGHT_X, CON.CONST_BOTTOM_RIGHT_Y), (0,0,255), 4)
	else:
		#cv2.imwrite("./data/"+str(temp)+'frame.png',frame)
		print("###############################################\n")
		print("Is Frame showing movement :::::::::::::::: ",r1)
		print("Is Flame content valid ::::::::::::::::::: ",r2)
		print("Is Flame size valid :::::::::::::::::::::: ",r3)
		print("\n###############################################\n")
		cv2.rectangle(frame, (CON.CONST_UPPER_LEFT_X, CON.CONST_UPPER_LEFT_Y), (CON.CONST_BOTTOM_RIGHT_X, CON.CONST_BOTTOM_RIGHT_Y), (0,255,0), 4)
	
	cv2.imshow('Flame Detection',frame)
	cv2.imshow('Motion of pixels',prev_bgr)
	
	if cv2.waitKey(1) & 0xFF == ord('q'):
		break


cap.release()
cv2.destroyAllWindows()

File: Cuda_Version/FlameDetection.cpp
#include <iostream>
#include "opencv2/opencv_modules.hpp"
#include <opencv2/core/core.hpp>
#include <opencv2/core/cuda.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core.hpp>
#include <opencv2/core/opengl.hpp>
#include <opencv2/cudacodec.hpp>
#include <opencv2/highgui.hpp>
#include "opencv2/opencv.hpp"

#include <string>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;
using namespace cv;
using namespace cv::cuda;

const int X_CORDINATE = 575;
const int Y_CORDINATE = 45;
const int WIDTH =  60;
const int HEIGHT = 50;

const int CONST_AREA_THRESHOLD = 30;
const int CONST_FLAME_THRESHOLD = 10;
const int THICKNESS = 2;

bool Flame_Area_Detection(const Mat &src){
    // Converting grayscale to black n white
    cv::Mat bw = src > .5;
    cv::blur( bw, bw, cv::Size(3,3) );
   
    int count_white = 0;
    for(int i =0 ; i< bw.rows; i++){
        for(int j =0 ; j< bw.cols; j++){
            if(bw.at<float>(i,j) != 0){
              count_white++;
            }
        }
    }

    float per_b_pixel =((float)count_white/(float)(bw.rows*bw.cols))* 100;
    if(per_b_pixel < CONST_AREA_THRESHOLD)
        return false;
    return true;
}

bool get_HSI(const Mat &pic){

    cv::Mat hsi(pic.rows, pic.cols, pic.type());
    int H = pic.rows;
    int W = pic.cols;
    bool grab = true;
    for (int j=0;j<H;j++)
    for (int i=0;i<W;i++) {

        double B =(double) pic.at<cv::Vec3b>(j,i)[0];
        double G =(double) pic.at<cv::Vec3b>(j,i)[1];
        double R =(double) pic.at<cv::Vec3b>(j,i)[2];
        double intensity = 0;
        double hue = 0;
        double saturation = 0;
    
        intensity = (double) (R + G + B) / (3.0);
        double tmp = min(R, min(G, B));

        saturation = 1 - 3*(tmp/(R + G + B));
        if(saturation < 0.00001){
            saturation = 0;
        }else if(saturation > 0.99999){
                saturation = 1;
        }

        double rb = R - B;
		double rg = R - G;
		double denom = sqrt(( rg * rg) + (rb*(G - B)));
		if( saturation != 0 and denom!=0){
            hue = 0.5 * (rg - rb) /sqrt(( rg * rg) + (rb*(G - B)));
			hue = acos(hue);
			if( B <= G){
                    hue = hue;
            }
			else{
                    hue = 360 -hue;
            }
        }		
		else{
                    hue = 180;
        }
        saturation = saturation * 100;
        hsi.at<cv::Vec3b>(j, i)[2] = hue;
        hsi.at<cv::Vec3b>(j, i)[1] = saturation;
        hsi.at<cv::Vec3b>(j, i)[0] = intensity;
        }

    int count = 0;
	for (int j=0;j< H;j++)
       for (int i=0;i<W;i++){

        double I =(double) hsi.at<cv::Vec3b>(j,i)[0];
        double S =(double) hsi.at<cv::Vec3b>(j,i)[1];
        double H =(double) hsi.at<cv::Vec3b>(j,i)[2];
		//Tune values of H,S,I as per demand 
		if (!((H<=120) and (S<=50) and (I>=180))){
                hsi.at<cv::Vec3b>(j, i)[2] = 0;
                hsi.at<cv::Vec3b>(j, i)[1] = 0;
                hsi.at<cv::Vec3b>(j, i)[0] = 0;
				count+=1;
        }
    }

	int size = (H * W);
	float flame_per = 100 - (count*100)/size;

	if (flame_per < CONST_FLAME_THRESHOLD){
              return false;
    }

    //cout<< flame_per<<endl;
    cv::imshow("HSI Format",hsi);
    return true;
}


bool Motion_Detection(const Mat &prev ,const  Mat &next) {
    Mat flow;
    cv::calcOpticalFlowFarneback(prev,next,flow, 0.5,3,15,3,5,1.2,0);
    cv::Mat xy[2]; //X,Y
    cv::split(flow, xy);

    //calculate angle and magnitude
    cv::Mat magnitude, angle;
    cv::cartToPolar(xy[0], xy[1], magnitude, angle, true);

    double mag_max;
    cv::minMaxLoc(magnitude, 0, &mag_max);
    Mat mag;
    
    normalize(magnitude, mag, 0, 255, NORM_MINMAX);
    cv::Mat _hsv[3], hsv;
    _hsv[0] = angle ; //((angle*180)/3.14)/2;
    _hsv[1] = cv::Mat::ones(angle.size(), CV_32F);
    _hsv[1] = 255;
    _hsv[2] = mag;
    cv::merge(_hsv, 3, hsv);

    //convert to BGR and show
    cv::Mat bgr;//CV_32FC3 matrix
    cuda::GpuMat hsv_cuda, bgr_cuda;
    hsv_cuda.upload(hsv);
    cuda::cvtColor(hsv_cuda, bgr_cuda, cv::COLOR_HSV2BGR);
    bgr_cuda.download(bgr);

    int count = 0;
    for(int i =0 ; i<bgr.size[0]; i++ ){
        for(int j =0 ; j<bgr.size[1]; j++ ){
            if(bgr.at<float>(i,j) != 0){
                    //cout<< bgr.at<Vec3b>(i,j)[0];
                    count++;
            }
        }
    }   
     int thresh_hold = (bgr.size[0] * bgr.size[1])/3;
     if(count >= thresh_hold){
         cv::imshow("optical flow", bgr);
         return (true) ;
     }
    cv::imshow("optical flow", bgr);
    return false;
}

int main() {

    //Reading videa frames from the dir.
    VideoCapture cap("./../Data/flame.mkv");
    if (!cap.isOpened())
        return -1;

    // Declaring Mat and GPU variables
    GpuMat gpu_frame;
    Mat frame;
    cv::Mat greyMat;
    cv::Mat prevFrame;
    GpuMat prev_frame;
    bool grabFrame = true;
    bool r1,r2,r3;

    cap >> frame;

    frame.convertTo(frame, CV_32F);
    gpu_frame.upload(frame);
    gpu_frame.convertTo(gpu_frame,CV_32F, 1.f/ 255);

    cv::cuda::GpuMat image(gpu_frame);
    cv::Rect regionOfInterest( X_CORDINATE ,Y_CORDINATE, WIDTH, HEIGHT);
    gpu_frame = image(regionOfInterest);

    cv::cuda::cvtColor(gpu_frame, prev_frame, CV_BGR2GRAY);
    prev_frame.download(prevFrame);

    while (grabFrame) {
        // Grab frame
        cap >> frame;
        Mat frame_HSI = frame;
        frame.convertTo(frame, CV_32F);
        gpu_frame.upload(frame);

        gpu_frame.convertTo(gpu_frame,CV_32F, 1.f/ 255);
        cv::cuda::GpuMat image(gpu_frame);
        cv::Rect regionOfInterest( X_CORDINATE ,Y_CORDINATE, WIDTH, HEIGHT);
        gpu_frame = image(regionOfInterest);

        GpuMat next_frame;
        cv::cuda::cvtColor(gpu_frame, next_frame, CV_BGR2GRAY);
        next_frame.download(greyMat);
        Mat croppedImg;
        frame_HSI(cv::Rect(X_CORDINATE,Y_CORDINATE,WIDTH,HEIGHT)).copyTo(croppedImg);

        r1 = Motion_Detection( prevFrame , greyMat);
        r2 = get_HSI(croppedImg);
        r3 = Flame_Area_Detection(greyMat);

        //r1 && r2 && r3
        if(r1 && r2 && r3){
            cv::rectangle(frame_HSI ,regionOfInterest, cv::Scalar(0, 0, 255),THICKNESS);
        }else{
            	cout<<"###############################################"<<endl;
		        cout<<"Is Frame showing movement :::::::::::::::: "<<r1<<endl;
		        cout<<"Is Flame content valid ::::::::::::::::::: "<<r2<<endl;
		        cout<<"Is Flame size valid :::::::::::::::::::::: "<<r3<<endl;
		        cout <<"###############################################"<<endl;
            cv::rectangle(frame_HSI ,regionOfInterest, cv::Scalar(0, 255, 0), THICKNESS);
        }

        prev_frame = next_frame;
        next_frame.download(prevFrame);

        if (frame.empty())
            {
                cout<<"Video Ended Successfully"<<endl;
                break;
            }
        imshow("Flame Detection", frame_HSI);
        if (cv::waitKey(30) >= 0)
            grabFrame = false;
    }

    cap.release();
    destroyAllWindows();
    return 0;
}

